<a href="https://colab.research.google.com/github/MuHyeonSon/algorithm/blob/main/codingtest_with_python/DFS%26BFS/%EC%97%B0%EA%B5%AC%EC%86%8C_%EC%88%98%ED%96%89%EC%8B%9C%EA%B0%84%EB%B9%84%EA%B5%90.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 교재 풀이와 나의 풀이의 수행시간 비교
## condingtest_with_python_part3_DFS/BFS_연구소

나의 풀이

In [1]:

# 나의 풀이 (주어진 풀이 시간 : 40분, 풀이 시간  60분 초 )
import copy
from collections import deque
import time

n, m = map(int, input().split())
Map = []

for _ in range(n):
    Map.append(list(map(int, input().split())))
## bfs로 모두 탐색하다가 2를 발견하면 인접노드(상하좌우)가 1인 곳을 찾는다.
## 2의 상하좌우인 곳 중 그 위치의 원소 값이 0인 곳은 모두 2로 갱신
## 새로 벽을 세울 위치를 3중 반복문 돌려서 모든 경우 다 탐색 해봐?
## 2의 인접노드를 볼거니까 BFS이용?
start_time = time.time()

max_value = 0

zero_list = []

temp_Map = copy.deepcopy(Map)

def bfs(temp_Map, start):
    #q = deque([start])
    a, b = start # 2 인 곳(바이러스가 있는 곳)
    # temp_Map[a][b] = -1 # 방문 표시
    # 상하좌우 이동방향 벡터
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    for i in range(4):
        nr = a + dr[i]
        nc = b + dc[i]
        if 0 <= nr < n and 0 <= nc < m and temp_Map[nr][nc] != 1:
            temp_Map[nr][nc] = 2
        
    

 
for i in range(n):
    for j in range(m):
        if Map[i][j] == 0:
            zero_list.append((i, j)) # 빈칸인 a행 b열을 빈칸 리스트에 추가

for i in range(len(zero_list)):
    for j in range(len(zero_list)):
        for k in range(len(zero_list)):
            if i == j or i == k or j == k:
                continue
            if temp_Map[zero_list[i][0]][zero_list[i][1]] != 0 or temp_Map[zero_list[j][0]][zero_list[j][1]] != 0 or temp_Map[zero_list[k][0]][zero_list[k][1]] != 0:
                continue
            else:
                temp_Map[zero_list[i][0]][zero_list[i][1]] = 1
                temp_Map[zero_list[j][0]][zero_list[j][1]] = 1
                temp_Map[zero_list[k][0]][zero_list[k][1]] = 1
                for row in range(n):
                    for column in range(m):
                        if temp_Map[row][column] == 2:
                            bfs(temp_Map, (row, column))
                count = 0
                for r in range(n):
                    for c in range(m):
                        if temp_Map[r][c] == 0:
                            count += 1
                max_value = max(max_value, count)
                temp_Map = copy.deepcopy(Map)

print(max_value)
                        
end_time = time.time()

print(f'수행시간 : {end_time - start_time}')            
                

7 7
2 0 0 0 1 1 0
0 0 1 0 1 2 0
0 1 1 0 1 0 0
0 1 0 0 0 0 0
0 0 0 0 0 1 1
0 1 0 0 0 0 0
0 1 0 0 0 0 0
27
수행시간 : 4.632873773574829


교재 풀이

In [2]:
import time
# 교재 풀이
n, m = map(int, input().split())
data = [] # 초기 맵 리스트
temp = [[0] * m for _ in range(n)]

for _ in range(n):
    data.append(list(map(int, input().split())))

start_time = time.time()

## 4가지 이동 방향에 대한 리스트
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]

result = 0

## 깊이 우선 탐색(DFS)를 이용해 각 바이러스가 사방으로 퍼지도록 하기
def virus(r, c):
    for i in range(4):
        nr = r + dr[i]
        nc = c + dc[i]
        # 상, 하, 좌, 우 중에서 바이러스가 퍼질 수 있는 경우
        if 0 <= nr < n and 0 <= nc < m:
            if temp[nr][nc] == 0:
                # 해당 위치에 바이러스 배치하고, 다시 재귀적으로 수행
                temp[nr][nc] = 2
                virus(nr, nc) 
    
## 현재 맵에서 안전 영역의 크기를 계산하는 메서드
def get_score():
    score = 0
    for i in range(n):
        for j in range(m):
            if temp[i][j] == 0:
                score += 1
    return score

## 깊이 우선 탐색(DFS)을 이용해 울타리를 설치하면서, 매번 안전 영역의 크기를 계산
def dfs(count):
    global result
    # 울타리가 3개 설치된 경우
    if count == 3:
        for i in range(n):
            for j in range(m):
                temp[i][j] = data[i][j]
        # 각 바이러스의 위치에서 전파 진행
        for i in range(n):
            for j in range(m):
                if temp[i][j] == 2:
                    virus(i, j)
        # 안전 영역의 최댓값 계산
        result = max(result, get_score())
        return
    # 빈 공간에 울타리 설치
    for i in range(n):
        for j in range(m):
            if data[i][j] == 0:
                data[i][j] = 1
                count += 1
                dfs(count)
                data[i][j] = 0
                count -= 1

dfs(0)
print(result)

end_time = time.time()

print(f'수행시간 : {end_time - start_time}')  

7 7
2 0 0 0 1 1 0
0 0 1 0 1 2 0
0 1 1 0 1 0 0
0 1 0 0 0 0 0
0 0 0 0 0 1 1
0 1 0 0 0 0 0
0 1 0 0 0 0 0
27
수행시간 : 4.966391086578369


###내가 비효율적으로 작성한 코드보다 교재 풀이 코드가 더 느린 것을 확인했는데 어떻게 해야하는지..

다른 테스트케이스도 비교해보자

나의 풀이 (test case2)

In [5]:

# 나의 풀이 (주어진 풀이 시간 : 40분, 풀이 시간  60분 초 )
import copy
from collections import deque
import time

n, m = map(int, input().split())
Map = []

for _ in range(n):
    Map.append(list(map(int, input().split())))
## bfs로 모두 탐색하다가 2를 발견하면 인접노드(상하좌우)가 1인 곳을 찾는다.
## 2의 상하좌우인 곳 중 그 위치의 원소 값이 0인 곳은 모두 2로 갱신
## 새로 벽을 세울 위치를 3중 반복문 돌려서 모든 경우 다 탐색 해봐?
## 2의 인접노드를 볼거니까 BFS이용?
start_time = time.time()

max_value = 0

zero_list = []

temp_Map = copy.deepcopy(Map)

def bfs(temp_Map, start):
    #q = deque([start])
    a, b = start # 2 인 곳(바이러스가 있는 곳)
    # temp_Map[a][b] = -1 # 방문 표시
    # 상하좌우 이동방향 벡터
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    for i in range(4):
        nr = a + dr[i]
        nc = b + dc[i]
        if 0 <= nr < n and 0 <= nc < m and temp_Map[nr][nc] != 1:
            temp_Map[nr][nc] = 2
        
    

 
for i in range(n):
    for j in range(m):
        if Map[i][j] == 0:
            zero_list.append((i, j)) # 빈칸인 a행 b열을 빈칸 리스트에 추가

for i in range(len(zero_list)):
    for j in range(len(zero_list)):
        for k in range(len(zero_list)):
            if i == j or i == k or j == k:
                continue
            if temp_Map[zero_list[i][0]][zero_list[i][1]] != 0 or temp_Map[zero_list[j][0]][zero_list[j][1]] != 0 or temp_Map[zero_list[k][0]][zero_list[k][1]] != 0:
                continue
            else:
                temp_Map[zero_list[i][0]][zero_list[i][1]] = 1
                temp_Map[zero_list[j][0]][zero_list[j][1]] = 1
                temp_Map[zero_list[k][0]][zero_list[k][1]] = 1
                for row in range(n):
                    for column in range(m):
                        if temp_Map[row][column] == 2:
                            bfs(temp_Map, (row, column))
                count = 0
                for r in range(n):
                    for c in range(m):
                        if temp_Map[r][c] == 0:
                            count += 1
                max_value = max(max_value, count)
                temp_Map = copy.deepcopy(Map)

print(max_value)
                        
end_time = time.time()

print(f'수행시간 : {end_time - start_time}')            
                

4 6
0 0 0 0 0 0
1 0 0 0 0 2
1 1 1 0 0 2
0 0 0 0 0 2
13
수행시간 : 0.17273497581481934


교재 풀이 (test case2)

In [6]:
import time
# 교재 풀이
n, m = map(int, input().split())
data = [] # 초기 맵 리스트
temp = [[0] * m for _ in range(n)]

for _ in range(n):
    data.append(list(map(int, input().split())))

start_time = time.time()

## 4가지 이동 방향에 대한 리스트
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]

result = 0

## 깊이 우선 탐색(DFS)를 이용해 각 바이러스가 사방으로 퍼지도록 하기
def virus(r, c):
    for i in range(4):
        nr = r + dr[i]
        nc = c + dc[i]
        # 상, 하, 좌, 우 중에서 바이러스가 퍼질 수 있는 경우
        if 0 <= nr < n and 0 <= nc < m:
            if temp[nr][nc] == 0:
                # 해당 위치에 바이러스 배치하고, 다시 재귀적으로 수행
                temp[nr][nc] = 2
                virus(nr, nc) 
    
## 현재 맵에서 안전 영역의 크기를 계산하는 메서드
def get_score():
    score = 0
    for i in range(n):
        for j in range(m):
            if temp[i][j] == 0:
                score += 1
    return score

## 깊이 우선 탐색(DFS)을 이용해 울타리를 설치하면서, 매번 안전 영역의 크기를 계산
def dfs(count):
    global result
    # 울타리가 3개 설치된 경우
    if count == 3:
        for i in range(n):
            for j in range(m):
                temp[i][j] = data[i][j]
        # 각 바이러스의 위치에서 전파 진행
        for i in range(n):
            for j in range(m):
                if temp[i][j] == 2:
                    virus(i, j)
        # 안전 영역의 최댓값 계산
        result = max(result, get_score())
        return
    # 빈 공간에 울타리 설치
    for i in range(n):
        for j in range(m):
            if data[i][j] == 0:
                data[i][j] = 1
                count += 1
                dfs(count)
                data[i][j] = 0
                count -= 1

dfs(0)
print(result)

end_time = time.time()

print(f'수행시간 : {end_time - start_time}')  

4 6
0 0 0 0 0 0
1 0 0 0 0 2
1 1 1 0 0 2
0 0 0 0 0 2
9
수행시간 : 0.18733453750610352


나의 풀이 (test case3)

In [7]:

# 나의 풀이 (주어진 풀이 시간 : 40분, 풀이 시간  60분 초 )
import copy
from collections import deque
import time

n, m = map(int, input().split())
Map = []

for _ in range(n):
    Map.append(list(map(int, input().split())))
## bfs로 모두 탐색하다가 2를 발견하면 인접노드(상하좌우)가 1인 곳을 찾는다.
## 2의 상하좌우인 곳 중 그 위치의 원소 값이 0인 곳은 모두 2로 갱신
## 새로 벽을 세울 위치를 3중 반복문 돌려서 모든 경우 다 탐색 해봐?
## 2의 인접노드를 볼거니까 BFS이용?
start_time = time.time()

max_value = 0

zero_list = []

temp_Map = copy.deepcopy(Map)

def bfs(temp_Map, start):
    #q = deque([start])
    a, b = start # 2 인 곳(바이러스가 있는 곳)
    # temp_Map[a][b] = -1 # 방문 표시
    # 상하좌우 이동방향 벡터
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    for i in range(4):
        nr = a + dr[i]
        nc = b + dc[i]
        if 0 <= nr < n and 0 <= nc < m and temp_Map[nr][nc] != 1:
            temp_Map[nr][nc] = 2
        
    

 
for i in range(n):
    for j in range(m):
        if Map[i][j] == 0:
            zero_list.append((i, j)) # 빈칸인 a행 b열을 빈칸 리스트에 추가

for i in range(len(zero_list)):
    for j in range(len(zero_list)):
        for k in range(len(zero_list)):
            if i == j or i == k or j == k:
                continue
            if temp_Map[zero_list[i][0]][zero_list[i][1]] != 0 or temp_Map[zero_list[j][0]][zero_list[j][1]] != 0 or temp_Map[zero_list[k][0]][zero_list[k][1]] != 0:
                continue
            else:
                temp_Map[zero_list[i][0]][zero_list[i][1]] = 1
                temp_Map[zero_list[j][0]][zero_list[j][1]] = 1
                temp_Map[zero_list[k][0]][zero_list[k][1]] = 1
                for row in range(n):
                    for column in range(m):
                        if temp_Map[row][column] == 2:
                            bfs(temp_Map, (row, column))
                count = 0
                for r in range(n):
                    for c in range(m):
                        if temp_Map[r][c] == 0:
                            count += 1
                max_value = max(max_value, count)
                temp_Map = copy.deepcopy(Map)

print(max_value)
                        
end_time = time.time()

print(f'수행시간 : {end_time - start_time}')            
                

8 8
2 0 0 0 0 0 0 2
2 0 0 0 0 0 0 2
2 0 0 0 0 0 0 2
2 0 0 0 0 0 0 2
2 0 0 0 0 0 0 2
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
8
수행시간 : 31.011623859405518


교재 풀이 (test case3)

In [10]:
import time
# 교재 풀이
n, m = map(int, input().split())
data = [] # 초기 맵 리스트
temp = [[0] * m for _ in range(n)]

for _ in range(n):
    data.append(list(map(int, input().split())))

start_time = time.time()

## 4가지 이동 방향에 대한 리스트
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]

result = 0

## 깊이 우선 탐색(DFS)를 이용해 각 바이러스가 사방으로 퍼지도록 하기
def virus(r, c):
    for i in range(4):
        nr = r + dr[i]
        nc = c + dc[i]
        # 상, 하, 좌, 우 중에서 바이러스가 퍼질 수 있는 경우
        if 0 <= nr < n and 0 <= nc < m:
            if temp[nr][nc] == 0:
                # 해당 위치에 바이러스 배치하고, 다시 재귀적으로 수행
                temp[nr][nc] = 2
                virus(nr, nc) 
    
## 현재 맵에서 안전 영역의 크기를 계산하는 메서드
def get_score():
    score = 0
    for i in range(n):
        for j in range(m):
            if temp[i][j] == 0:
                score += 1
    return score

## 깊이 우선 탐색(DFS)을 이용해 울타리를 설치하면서, 매번 안전 영역의 크기를 계산
def dfs(count):
    global result
    # 울타리가 3개 설치된 경우
    if count == 3:
        for i in range(n):
            for j in range(m):
                temp[i][j] = data[i][j]
        # 각 바이러스의 위치에서 전파 진행
        for i in range(n):
            for j in range(m):
                if temp[i][j] == 2:
                    virus(i, j)
        # 안전 영역의 최댓값 계산
        result = max(result, get_score())
        return
    # 빈 공간에 울타리 설치
    for i in range(n):
        for j in range(m):
            if data[i][j] == 0:
                data[i][j] = 1
                count += 1
                dfs(count)
                data[i][j] = 0
                count -= 1

dfs(0)
print(result)

end_time = time.time()

print(f'수행시간 : {end_time - start_time}')  

8 8
2 0 0 0 0 0 0 2
2 0 0 0 0 0 0 2
2 0 0 0 0 0 0 2
2 0 0 0 0 0 0 2
2 0 0 0 0 0 0 2
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
3
수행시간 : 31.01591444015503


나의 풀이-----------------------------------------교재 풀이

수행시간 : 4.632873773574829
수행시간 : 4.966391086578369

수행시간 : 0.17273497581481934
수행시간 : 0.18733453750610352

수행시간 : 31.011623859405518
수행시간 : 31.01591444015503

위와 같이 모든 테스트케이스에서 나의 풀이가 교재 풀이보다 더 빠른 성능을
보였다.. 어떤 것이 잘 못된 것인지 모르겠다. 분명 vscode로 돌렸을 때도
이러한 현상이 있었고, 수행시간 자체도 더 걸려서 코랩환경에서 돌려본 것인데,
이렇게 되면 둘 다 수행시간 2초를 넘어가기 때문에 문제가 있다.. 